snf-cyclades-app/synnefo/api/keypairs.py

```diff
@@ -0,0 +1,160 @@
+
+def demux(request):
+    if request.method == 'GET':
+        return list_keypairs(request)
+    elif request.method == 'POST':
+        return create_new_keypair(request)
+    else:
+        return api.api_method_not_allowed(request,
+                                          allowed_methods=['GET', 'POST'])
+
+
+def keypair_demux(request, keypair_name):
+    if request.method == 'GET':
+        return get_keypair(request, keypair_name)
+    if request.method == 'DELETE':
+        return delete_keypair(request, keypair_name)
+    else:
+        return api.api_method_not_allowed(request,
+                                          allowed_methods=['GET', 'DELETE'])
+
+
+@api.api_method(http_method='GET', user_required=True, logger=log)
+def list_keypairs(request):
+    active_keypairs = PublicKeyPair.objects.filter(user=request.user_uniq).all()
+    keypairs = [keypair_to_dict(keypair) for keypair in
+                active_keypairs.order_by('name')]
+    data = json.dumps({'keypairs': keypairs})
+    return HttpResponse(data, status=200)
+
+
+@api.api_method(http_method='POST', user_required=True, logger=log)
+def create_new_keypair(request):
+
+    get_user(request, settings.ASTAKOS_AUTH_URL)
+
+    if PublicKeyPair.user_limit_exceeded(request.user_uniq):
+        return HttpResponseServerError("SSH keys limit exceeded")
+
+    req = utils.get_json_body(request)
+    try:
+        keypair = req['keypair']
+        assert(isinstance(req, dict))
+        name = keypair['name']
+    except (KeyError, AssertionError):
+        raise faults.BadRequest('Malformed request.')
+
+    data_dict = {'name': name}
+    try:
+        # If the public_key is provided  and the corresponding
+        # keypair already exists update the public key
+        keypair_obj = util.get_keypair(name, request.user_uniq)
+    except PublicKeyPair.DoesNotExist:
+        keypair_obj = PublicKeyPair(
+                name=name, user=request.user_uniq)
+
+    try:
+        public_key = keypair['public_key']
+        keypair_obj.content = public_key
+    except KeyError:
+        # If the public_key field is ommited, generate a new
+        # keypair and return both the private and the public key
+        if not SUPPORT_GENERATE_KEYS:
+            raise Exception(
+                "Application does not support ssh keys generation")
+
+        new_keypair = generate_keypair()
+        keypair_obj.content = new_keypair['public']
+        data_dict['private_key'] = new_keypair['private']
+
+    keypair_obj.save()
+
+    data_dict['fingerprint'] = keypair_obj.fingerprint
+    data_dict['public_key'] = keypair_obj.content
+    data = json.dumps(data_dict)
+
+    return HttpResponse(data, status=201)
+
+
+@api.api_method(http_method='GET', user_required=True, logger=log)
+def get_keypair(request, keypair_name):
+    try:
+        get_user(request, settings.ASTAKOS_AUTH_URL)
+        keypair = util.get_keypair(keypair_name, request.user_uniq)
+    except PublicKeyPair.DoesNotExist:
```

Comment:
This handing is not needed anymore if we change `util.get_keypair` accordingly, see below. The `@api.api_method` will handle the `faults.ItemNotFound` exception, so we will not need to handle it ourselves.